<p>
<style type="text/css">
a { text-decoration:underline; }
a:hover { text-decoration:underline; }
a:active { text-decoration:underline; }</style>
</p><h2><span style="font-size:22px;">Download</span></h2><p>Archive: <a href="https://dl.dropboxusercontent.com/u/770999/az/orthofinder.zip">https://dl.dropboxusercontent.com/u/770999/az/orthofinder.zip</a></p><p>Repository:&nbsp;<a href="https://github.com/vladsaveliev/az_orthofinder">https://github.com/vladsaveliev/az_orthofinder</a></p><div>&nbsp;</div><h2><span style="font-size: 22px;">Installation</span></h2><div><span style="font-size:14px;">Just extract the archive. You will find scenario_1.py, scenario_2.py and test_input inside the extracted folder.</span></div><div><span style="font-size:14px;"><em>Note</em>: you will need some third-party software to be installed on your system for running the tool. See section System Requirements for details.</span></div><h2>&nbsp;</h2><h2><span style="font-size: 22px;">Scenario 1</span></h2><div><div>The <span style="font-family:courier new,courier,monospace;">scenario_1.py</span> is aimed to initialize a database of orthologous groups. It generates the resulting <span style="font-family:courier new,courier,monospace;">orthogroups.tsv</span>, and for further extention it also produces the following intermediate results for the second scenario:&nbsp;</div><div>— the <span style="font-family:courier new,courier,monospace;">proteomes</span> directory with correctly adjusted proteomes,&nbsp;</div><div>— the <span style="font-family:courier new,courier,monospace;">annotations</span> directory with GB files from NCBI,&nbsp;</div><div>— and the <span style="font-family:courier new,courier,monospace;">intermediate/blasted.tsv</span>&nbsp;file.</div><div>&nbsp;</div><div><em>Usage examples:</em></div><div><strong>1. </strong>Fasta-proteins (optionally with annotations from prodigal). See example in <span style="font-family:courier new,courier,monospace;">test_input/proteins</span>. Filenames will be taken as taxon codes. Uses Internet to download GB annotations; it the Internet is off, a short version of output will be produced, containing only taxon|protein ids.</div><p style="margin-left: 40px;"><span style="font-size:14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_1.py --proteomes </span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">test_input/proteins</span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 14px;">&nbsp;-o output_test_proteomes</span></p><p><strong>2. </strong>GB-annotations (Example:&nbsp;<span style="font-family:courier new,courier,monospace;">test_input/gbs<span style="font-family:arial,helvetica,sans-serif;">)</span></span>.</p><p style="margin-left: 40px;"><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">./</span><span style="font-size: 14px; background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace;">scenario_1.py --gbs test_input/gbs</span><span style="font-size: 14px; background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace;">&nbsp;-o output_test_gbs</span></p><p><strong>3. </strong>A list of reference ids. (Example:&nbsp;<span style="font-family:courier new,courier,monospace;">test_input/ids.txt<span style="font-family:arial,helvetica,sans-serif;">)</span></span><span style="font-family:arial,helvetica,sans-serif;">. </span>The tool will download references from Genbank. This step requires an Internet connection.</p><div style="margin-left: 40px;"><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">./</span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 14px;">scenario_1.py -i test_input/ids.txt -o output_test_ids</span></div><div>&nbsp;</div><div><div><strong>4. </strong>A&nbsp;list of species names (Example:&nbsp;<span style="font-family:courier new,courier,monospace;">test_input/species.txt</span>). In this case, the tool will search the NCBI server using the following query (considering a species name is Escherichia coli):</div><div><a href="http://www.ncbi.nlm.nih.gov/nuccore/?term=(Escherichia+coli%5BOrganism%5D+AND+(complete+genome%5BTitle%5D+OR+complete+sequence%5BTitle%5D))+NOT+(partial%5BTitle%5D)">Escherichia coli[Organism] AND (complete genome[Title] OR complete sequence[Title]))&nbsp;NOT (partial[Title])</a></div><div>Note that you need to specify full species name like Escherichia coli (not E. coli).</div><div>&nbsp;</div><div>Particularly, we processed E. coli and K. pneumoniae this way. I used the files with species list (see the test_input directory):</div><div>kpneumonia_list.txt with the following single line inside:</div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">Klebsiella pneumoniae</span></div><div>The command line I used:&nbsp;</div><p style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_1</span></span><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">.py&nbsp;</span></span></span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 14px;">--species test_input/kpneumonia_list.txt</span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">&nbsp;-o kpneumonia</span></p><div><div>&nbsp;</div><div>ecoli_list.txt contating the following:</div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">Escherichia coli</span></div><div>Command line:&nbsp;</div></div><div style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_1</span></span><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">.py&nbsp;</span></span></span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 14px;">--species test_input/ecoli_list.txt</span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">&nbsp;-o ecoli</span></div><div style="margin-left: 40px;">&nbsp;</div><div style="margin-left: 40px;">&nbsp;</div><div><h2><span style="font-size: 22px;">Scenario 2</span></h2><div><div>The <span style="font-family:courier new,courier,monospace;">scenario_2.py</span> script is meant to extend orthogroups. It is required to run on a direcotory produced by the <span style="font-family:courier new,courier,monospace;">scenario_1.py</span> script, since it extends it's <em>blasted.tsv</em>, and also reuses <em>proteomes</em> and GB&nbsp;<em>annotations</em>.&nbsp;</div><div><div>The reason file with blast results is used is that the all-against-all blast process is the most time consuming step. Basically, the blast results can be stored between runs in a database table; nevertheless, we desided to not to rely here on SQL, because we think it is going to be more clear for users:</div><div>1. The users don't have to remember names of their talbes so they won't overwrite or damage something important;</div><div>2. If the users want to send their results to somebody, they can just send the files. It would be not so obvious in case of MySQL.</div><div>3. The database can be cleaned up after any tool usage.</div><div>&nbsp;</div></div><div>There are 2 possible types of scenario_2 workflow depending on input.</div><div>1. Input is a list of reference IDs / GIs / organism names / GB annotations files.</div><div>In this case, the resulted intermediate files and othrogroups.tsv will be the same if you run the scenario_1.py on a larger input data.</div><div>&nbsp;</div><div>2. Input are assemblies or proteomes generated by Prodigal (in case of assemblies, proteomes will be generated automatically with Prodigal). After running scenario_2 on this input, <span style="font-family:courier new,courier,monospace;">orthogroups.tsv</span> is generated. Then each orthogroup is processed basing on it's kind:</div><div style="margin-left: 40px;">1. A group that contain only annotated genes is not processed any more.</div><div style="margin-left: 40px;">2. A group that contains both annotated and unknown genes is not processed as well, since unknown genes can be possibly curated based on annotated ones.</div><div style="margin-left: 40px;">3. If a group contains only unknown genes, it will saved into fasta file inside the blasted_singletones directory. Then, for each group, one of the proteins will be blasted against the public NCBI database (a local database can be also provided with the <span style="font-family:courier new,courier,monospace;">--blast-db</span> option; otherwise, an internet connection will be used).</div><div style="margin-left: 40px;">For each group, and XML file with blast results will be generated; the best hits will be printed to output.</div><div>&nbsp;</div></div><div><div><em>Usage examples:</em></div></div><div>Appending additional list of files (fasta, gb) to an existed output after scenario 1.</div></div></div><p style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_2.py&nbsp;</span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">output_test_proteomes</span><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">&nbsp;--proteomes </span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">test_input/new_proteins</span></p><p>You can pass assemblies instead, in this case they will be automatically annotated with Prodigal.</p><p style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_2.py&nbsp;</span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">output_test_proteomes</span><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">&nbsp;--assemblies&nbsp;</span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">test_input/assemblies</span></p><p>Or a list of reference ids (accession numbers of gi):</p><p style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_2.py&nbsp;</span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">output_test_proteomes</span><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">&nbsp;--ids&nbsp;</span></span></span><span style="font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px; background-color: rgb(211, 211, 211);">test_input/new_ids.txt</span></p><p>Exiting directory must contain an <span style="font-family: 'courier new', courier, monospace;">intermediate</span>&nbsp;subdirectory with a <span style="font-family: 'courier new', courier, monospace;">blasted.tsv</span> file and <span style="font-family: 'courier new', courier, monospace;">proteomes</span>&nbsp;folder from a scenario_1 run.</p><p>&nbsp;</p><h3><span style="font-size: 18px;">Starting from a step</span></h3><div>There is an optional command-line argument &nbsp;<span style="font-family: 'courier new', courier, monospace;">--start-from</span>. It is used to skip several steps of the pipeline and run right from the step specified. You can take step names from log.txt in the results folder.</div><div style="margin-left: 40px;"><span style="font-size:14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color:#d3d3d3;">./scenario_1</span></span><span style="font-family: 'courier new', courier, monospace;"><span style="background-color:#d3d3d3;">.py -o&nbsp;</span></span></span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">output_test_proteomes</span><span style="font-size: 14px; font-family: 'courier new', courier, monospace;"><span style="background-color:#d3d3d3;">&nbsp;</span></span><strong style="font-size: 14px; font-family: 'courier new', courier, monospace;"><span style="background-color:#d3d3d3;">--start-from&nbsp;</span></strong><span style="font-size: 14px; font-family: 'courier new', courier, monospace;"><span style="background-color:#d3d3d3;">"Parsing blast results"</span></span></div><div style="margin-left: 40px;">or</div><div style="margin-left: 40px;"><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">./scenario_1</span></span><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">.py</span></span></span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">&nbsp;-o</span><span style="font-size: 14px;"><span style="font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">&nbsp;</span></span></span><span style="background-color: rgb(211, 211, 211); font-family: 'courier new', courier, monospace; font-size: 13.63636302947998px;">output_test_proteomes</span><span style="font-size: 14px; font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">&nbsp;</span></span><strong style="font-size: 14px; font-family: 'courier new', courier, monospace;"><span style="background-color: rgb(211, 211, 211);">--start-from 7</span></strong></div><div style="margin-left: 40px;">&nbsp;</div><div><div><span style="font-style: italic;">scenario_1.py steps:</span></div><div style="margin-left: 40px;"><div>1. Preparing proteomes and annotations</div><div>2. Filtering proteomes</div><div>3. Making blast database</div><div>4. Blasting</div><div>5. Parsing blast results</div><div>6. Cleaning database</div><div>7. Installing schema</div><div>8. Loading blast results into the database</div><div>9. Finding pairs</div><div>10. Dump pairs files</div><div>11. MCL</div><div>12. Saving orthogroups&nbsp;</div></div><div style="margin-left: 40px;">&nbsp;</div></div><div><em>scenario_2.py steps:</em></div><div style="margin-left: 40px;">1. Preparing imput</div><div style="margin-left: 40px;">2. Filtering new proteomes</div><div style="margin-left: 40px;">3. Filtering proteomes</div><div style="margin-left: 40px;">4. Making blast database</div><div style="margin-left: 40px;">5. Blasting</div><div style="margin-left: 40px;">6. Parsing blast results</div><div style="margin-left: 40px;">7. Cleaning database</div><div style="margin-left: 40px;">8. Installing schema</div><div style="margin-left: 40px;">9. Loading blast results into the database</div><div style="margin-left: 40px;">10. Finding pairs</div><div style="margin-left: 40px;">11.&nbsp;Dumping pairs files</div><div style="margin-left: 40px;">12.&nbsp;MCL</div><div style="margin-left: 40px;">13.&nbsp;Saving orthogroups</div><div style="margin-left: 40px;">14.&nbsp;Blasting singletones</div><div style="margin-left: 40px;">&nbsp;</div><div><strong style="font-size: 1.2em;">Fine tuning</strong></div><div><div><div><span style="font-family: 'courier new', courier, monospace;">--prot-id-field &nbsp; &nbsp;&nbsp;</span>Fields are separated by either a bar or a space. For example, with --prot-id-filed 1 fasta ids like &gt;NC_005816.1|NP_995567.1 will lead to the protein id NP_995567.1</div><div>&nbsp;</div><div><span style="font-family: 'courier new', courier, monospace;">--min-length &nbsp; &nbsp; &nbsp; &nbsp;</span>Minimum allowed length of proteins (default: 10)</div></div><div><span style="font-family: 'courier new', courier, monospace;">--max-percent_stop &nbsp;</span>Maximum percent stop codons (default: 20)</div><div><span style="font-family: 'courier new', courier, monospace;">--evalue &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span>Blast e-value (default: 1e-5)</div><span style="font-family: 'courier new', courier, monospace;">-t &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span>Threads number (default: 30)</div></div><div>&nbsp;</div><div>&nbsp;</div><h2><span style="font-size:22px;"><strong style="font-size: 1.2em;">System Requirements</strong></span></h2><div><span id="cke_bm_90E" style="display: none;">&nbsp;</span>The tool needs the following software installed on your system:</div><ul><li>python 2.7</li><li>blast</li><li>mysql</li><li>mysql perl modules (note that src/mysql.cnf is a path from the root of the tool (az_orthofinder/src/mysql.cnf)):&nbsp;</li></ul><div><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 40px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">$ perl -MCPAN -e shell</p><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 40px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">cpan&gt; o conf makepl_arg "mysql_config=src/mysql.cnf"</p><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 40px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">cpan&gt; install Data::Dumper</p><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 40px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">cpan&gt; install DBI</p><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 40px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">cpan&gt; force install DBD::mysql</p><p style="margin-right: 0px; margin-bottom: 0.5em; margin-left: 0px; line-height: 19.200000762939453px; font-family: sans-serif; font-size: 12.727272033691406px;">&nbsp;</p></div><div>The tool also requires a MySQL user orthomcl with password 1234, and a database orthomcl with all privileges granted to that user. It can be achieved in the following way:</div><div><span style="font-family: 'courier new', courier, monospace;">$ mysqld --port=3307 &amp; &nbsp; (in case if the mysql server is not running)</span></div><div><span style="font-family: 'courier new', courier, monospace;">$ mysql -u root -p</span></div><div><span style="font-family: 'courier new', courier, monospace;">mysql&gt; CREATE DATABASE orthomcl;</span></div><div><span style="font-family: 'courier new', courier, monospace;">mysql&gt; GRANT SELECT, INSERT, UPDATE, DELETE, CREATE VIEW, CREATE, INDEX, DROP on orthomcl.* TO orthomcl@localhost;</span></div><div><span style="font-family: 'courier new', courier, monospace;">mysql&gt; set password for orthomcl@localhost = password("1234");</span></div><div>&nbsp;</div><div>If you have any problems when setting and running the mysql server, please, let us know immediately.</div><div>&nbsp;</div><div>&nbsp;</div><h2><span style="font-size:22px;"><span id="cke_bm_89S" style="display: none;">&nbsp;</span>Changes</span></h2><div><strong>Jan 29</strong></div><div>Fix: Correct running using a species list:</div><div>- includes plasmids</div><div>- takes first 100k records (not 20)</div><div>- fixed a bug that can cause an exception.</div><div>&nbsp;</div><div>Running using the --species option. &nbsp;</div><div>Considering a species.txt The tool searches Entrez</div><div>&nbsp;</div><div>&nbsp;</div><div><strong>Jan 28</strong></div><div>Scneario 2 can take a list of reference IDs: it downloads annotations from Genbank and appends proteins to existed blast results.</div><div>Usage:</div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">python scenario_2.py prev_dir --ids ids.txt</span></div><div>&nbsp;</div><div>As well as for the scenario 1, ids.txt can contain ranges: for example,&nbsp;</div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">CP003223:CP003228</span></div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">CP003230</span></div><div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">CP003232</span></div><div>&nbsp;</div><div><strong>Jan 24</strong></div><div>Scenario 2 searches singletone proteins against ref seq and outputs best protein sequences.</div><div>produces&nbsp;<span style="font-family:courier new,courier,monospace;">blasted_singletones </span>directory contains xml output of blast for each new protein, and txt file with a readable list of hits.</div><div>&nbsp;</div><div><strong>Jan 17</strong></div><div>1. Output format:</div><div><div style="margin-left: 40px;">1<span class="Apple-tab-span" style="white-space: pre;"> </span>Escherichia coli 55989<span class="Apple-tab-span" style="white-space: pre;"> </span>NC_011748.1<span class="Apple-tab-span" style="white-space: pre;"> </span>YP_002401224.1<span class="Apple-tab-span" style="white-space: pre;"> </span>EC55989_0084<span class="Apple-tab-span" style="white-space: pre;"> </span>murD<span class="Apple-tab-span" style="white-space: pre;"> </span>UDP-N-acetylmuramoyl-L-alanyl-D-glutamate synthetase<span class="Apple-tab-span" style="white-space: pre;"> </span></div><div><div style="margin-left: 40px;">1<span class="Apple-tab-span" style="white-space: pre;"> </span>Escherichia coli 55989<span class="Apple-tab-span" style="white-space: pre;"> </span>NC_011748.1<span class="Apple-tab-span" style="white-space: pre;"> </span>YP_002401227.1<span class="Apple-tab-span" style="white-space: pre;"> </span>EC55989_0087<span class="Apple-tab-span" style="white-space: pre;"> </span>murC<span class="Apple-tab-span" style="white-space: pre;"> </span>UDP-N-acetylmuramate--L-alanine ligase<span class="Apple-tab-span" style="white-space: pre;"> </span></div><div style="margin-left: 40px;">2<span class="Apple-tab-span" style="white-space: pre;"> </span>Escherichia coli 536<span class="Apple-tab-span" style="white-space: pre;"> </span>NC_008253.1<span class="Apple-tab-span" style="white-space: pre;"> </span>YP_667942.1<span class="Apple-tab-span" style="white-space: pre;"> </span>ECP_0001<span class="Apple-tab-span" style="white-space: pre;"> </span>NA<span class="Apple-tab-span" style="white-space: pre;"> </span>thr operon leader peptide<span class="Apple-tab-span" style="white-space: pre;"> </span></div><div style="margin-left: 40px;">2<span class="Apple-tab-span" style="white-space: pre;"> </span>Escherichia coli 55989<span class="Apple-tab-span" style="white-space: pre;"> </span>NC_011748.1<span class="Apple-tab-span" style="white-space: pre;"> </span>YP_002401144.1<span class="Apple-tab-span" style="white-space: pre;"> </span>EC55989_5066<span class="Apple-tab-span" style="white-space: pre;"> </span>thrL<span class="Apple-tab-span" style="white-space: pre;"> </span>thr operon leader peptide<span class="Apple-tab-span" style="white-space: pre;"> </span></div></div><div>&nbsp;</div></div><div>2. Logging date and time.</div><div>3. Fixed the issue with the database locks when too many records occur in a table.&nbsp;</div><div>4. Scenario 2 now accepts either assemblies (flag --assemblies), or proteins (flag --proteomes). In first case, it will run prodigal to find proteomes.</div><div>5. Scenario 2 saves separately those proteins that were not found separately in singletones directory.</div><div>&nbsp;</div><div><strong>Jan 16</strong></div><div>1. MCL is embeded into the tool and installed automatically.</div><div>2. Perl modules are also installed, by need sudo: installation without sudo is in progress.</div><div>&nbsp;</div><div><strong>Jan 14</strong></div><div><em>Fixes</em></div><div>1.&nbsp;Fixed "Saving orthogroups" step for scenario_1</div><div>&nbsp;</div><div><strong>Jan 13</strong></div><div>1. Checking the internet connection. If the test is failed and no annotations were provided manually, producing only short version of groups.txt.</div><div>2. When running on a directory with existing results, warn the user. An option <code>--overwrite</code>&nbsp;is used to suppress warning. Also, when using <code>--start-from, overwrite by default.</code></div><div>&nbsp;</div><div><em>Fixes</em></div><div>1. Taxon ids were retrived incorrectrly from proteomes names when contained a dot. Fixed this ussue.</div><div>2. Scenario_2.py: prodigal generated incorrect gbk file, that coused failure in reading genes. Now reading genes from the proteins .fasta file.</div><div style="margin-left: 40px;">&nbsp;</div><div><div><strong>Jan 10</strong></div><div>1. Default program argument now is directory,&nbsp;that contains either proteins fasta files (names of the files are taken as taxon codes), or genbank annotation files.</div><div>2. Optional arguments introduced for a fine adjustment of the process:&nbsp;--min-length,&nbsp;--max-percent_stop and --evalue.</div><div>3. New script, scenario_2.py. It takes an assembly and a directory, finds genes in the assembly, and then searches for orthologs using existing blast results.</div><div>&nbsp; &nbsp; Usage:</div><div style="margin-left: 40px;"><span style="font-family: 'courier new', courier, monospace;">python&nbsp;</span><span style="font-family: 'courier new', courier, monospace;">scenario_2</span><span style="font-family: 'courier new', courier, monospace;">.py directory assembly.fasta</span></div></div><div>&nbsp;</div><div><div><strong>Jan 6</strong></div><div>1. There are primary scripts now:&nbsp;<span style="font-family: 'courier new', courier, monospace;">scenario_1.py</span>&nbsp;and&nbsp;<span style="font-family: 'courier new', courier, monospace;">scenario_2.py</span>: first one is for annotating annotations massively, the second is for processing new assemblies (in progress).</div><div>2.&nbsp;<span style="font-family: 'courier new', courier, monospace;">scenario_2.py</span>&nbsp;now generates&nbsp;<span style="font-family: 'courier new', courier, monospace;">orthogroups.tsv</span>&nbsp;with tab-separated values:</div><div><div style="margin-left: 40px;">Orthogroup 1</div><div style="margin-left: 40px;">Group# Strain Name &nbsp; &nbsp;GI &nbsp; &nbsp; Gene &nbsp; &nbsp;Locus &nbsp; &nbsp;Product &nbsp; &nbsp; Description &nbsp; &nbsp;</div><div style="margin-left: 40px;"><div>2<span class="Apple-tab-span" style="white-space:pre"> </span>Escherichia coli 55989<span class="Apple-tab-span" style="white-space:pre"> </span>NC_011748.1<span class="Apple-tab-span" style="white-space:pre"> </span>YP_002401227.1<span class="Apple-tab-span" style="white-space:pre"> </span>EC55989_0087<span class="Apple-tab-span" style="white-space:pre"> </span>murC<span class="Apple-tab-span" style="white-space:pre"> </span>UDP-N-acetylmuramate--L-alanine ligase<span class="Apple-tab-span" style="white-space:pre"> </span></div><div>&nbsp;</div></div><div style="margin-left: 40px;">…</div><div style="margin-left: 40px;">Orthogroup 2</div><div style="margin-left: 40px;">Strain_name &nbsp; &nbsp;GI &nbsp; &nbsp; Gene &nbsp; &nbsp;Locus &nbsp; &nbsp;Product &nbsp; &nbsp; Description&nbsp;</div><div style="margin-left: 40px;">&nbsp;</div><div>it also generates&nbsp;<span style="font-family: 'courier new', courier, monospace;">orthogroups_nice.txt</span>&nbsp;with evenly aligned columns, which may be nicer to analyze manually rather that parse in code.</div><div>&nbsp;</div></div></div><div><strong>Dec 23</strong></div><div>1. Blast database now created instead of blasting fasta files. It allowed to run the tool in parallel (use -t N or --threads N).</div><div>2. Clearing mysql database step in overwrite mode (which is default).</div><div>3. Better scripts error code handling.</div><div>&nbsp;</div><div><em>Fixes</em></div><div>1. Failing if output directotory is in level above, i.e. -o ../results</div><div>2. Sometimes MySql exceptions codes are not resolved.</div><div>3. Checking existence of database tables for orthomclPairs, orthomclLoadBlast and orthomclDumpPairsFiles.</div>