<style type="text/css">
    a {
        text-decoration: underline;
    }

    a:hover {
        text-decoration: underline;
    }

    a:active {
        text-decoration: underline;
    }</style>
<div><span style="font-size:18px;">Latest version: <a href="https://dl.dropboxusercontent.com/u/770999/az/orthofinder.zip">orthofinder.zip</a></span>
</div>
<div>&nbsp;</div><h3><span style="font-size: 1.2em;">Changes</span></h3>
<div><strong>Dec 23</strong></div>
<div>1. Blast database now created instead of blasting fasta files. It allowed to run the tool in parallel (use -t N or
    --threads N).
</div>
<div>2. Clearing mysql database step in overwrite mode (which is default).</div>
<div>3. Better scripts error code handling.</div>
<div>&nbsp;</div>
<div><em>Fixes</em></div>
<div>1. Failing if output directotory is in level above, i.e. -o ../results</div>
<div>2. Sometimes MySql exceptions codes are not resolved.</div>
<div>3. Checking existence of database tables for orthomclPairs, orthomclLoadBlast and orthomclDumpPairsFiles.</div>
<div>&nbsp;</div><h3><strong style="font-size: 1.2em;">Requirements</strong></h3>
<div>The tool needs the following software installed on your system:</div>
<ul>
    <li>python 2.7</li>
    <li>blast</li>
    <li>mcl</li>
    <li>mysql</li>
</ul>
<div>&nbsp;</div>
<div>The tool also requires a MySQL user orthomcl with password 1234, and a database orthomcl with all privileges
    granted to that user. It can be achieved in the following way:
</div>
<div><span style="font-family:courier new,courier,monospace;">$ mysqld --port=3307 &amp; &nbsp; (in case if the mysql server is not running)</span>
</div>
<div><span style="font-family:courier new,courier,monospace;">$ mysql -u root -p</span></div>
<div><span style="font-family:courier new,courier,monospace;">mysql&gt; CREATE DATABASE orthomcl;</span></div>
<div><span style="font-family:courier new,courier,monospace;">mysql&gt; GRANT SELECT, INSERT, UPDATE, DELETE, CREATE VIEW, CREATE, INDEX, DROP on orthomcl.* TO orthomcl@localhost;</span>
</div>
<div><span style="font-family:courier new,courier,monospace;">mysql&gt; set password for orthomcl@localhost = password("1234");</span>
</div>
<div>&nbsp;</div>
<div>If you have any problems when setting and running the mysql server, please, let us know immediately.</div>
<div>&nbsp;</div><h3><strong style="font-size: 1.2em;">Usage</strong></h3>
<div>The tool can work with genbank annotation files. In this case you need to specify a path to a directory with .gb
    files using the <span style="font-family:courier new,courier,monospace;">--annotations-dir</span> option:
</div>
<div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">python find_orthologs.py --annotations-dir test_input/annotations -o results</span>
</div>
<div>&nbsp;</div>
<div>Alternatively, the tool accepts a list of reference ids with the <span
        style="font-family:courier new,courier,monospace;">--ids-file</span> option (see test_input/ids.txt for
    example). In this case, the tool will download references from genbank.
</div>
<div style="margin-left: 40px;"><span style="font-family:courier new,courier,monospace;">python find_orthologs.py --ids-file test_input/ids.txt -o results</span>
</div>
<div>&nbsp;</div>
<div>Or you can provide a list of species names (see example at test_input/species.txt). In this case, the tool will
    search the NCBI FTP server: ftp://ftp.ncbi.nih.gov/genomes/Bacteria/. Particularly, we processed E. coli and K.
    pneumoniae this way.
</div>
<div style="margin-left: 40px;"><span style="font-family: 'courier new', courier, monospace;">python find_orthologs.py --species-file test_input/species.txt -o results</span>
</div>
<div>&nbsp;</div>
<div>There is an optional command-line argument <span
        style="font-family:courier new,courier,monospace;">--start-from</span>. It is used to skip several steps of the
    pipeline and run right from the step specified. You can take step names from log.txt in the results folder.
</div>
<div style="margin-left: 40px;"><span style="font-family: 'courier new', courier, monospace;">python find_orthologs.py --ids-file test_input/ids.txt -o results </span><strong
        style="font-family: 'courier new', courier, monospace;">--start-from </strong><span
        style="font-family: 'courier new', courier, monospace;">"Parsing blast results"</span></div>
<div>&nbsp;</div>
<div>The tool can run in parallel:</div>
<div style="margin-left: 40px;"><span style="font-family: 'courier new', courier, monospace;">python find_orthologs.py --ids-file test_input/ids.txt -o results </span><strong
        style="font-family: 'courier new', courier, monospace;">-t </strong><span
        style="font-family: 'courier new', courier, monospace;">4</span></div>
<div>&nbsp;</div>